saleor/checkout/fetch.py

```diff
@@ -50,6 +65,157 @@ def get_customer_email(self) -> str:
+        return False
+
+    def update_channel_listings(self, checkout_info: "CheckoutInfo") -> None:
+        checkout_info.shipping_method_channel_listings = None
+
+
+@dataclass(frozen=True)
+class ShippingMethodInfo(DeliveryMethodBase):
+    delivery_method: "ShippingMethod"
+    shipping_address: Optional["Address"]
+    order_key: str = "shipping_method"
+
+    @property
+    def delivery_method_name(self) -> Dict[str, Optional[str]]:
+        return {"shipping_method_name": smart_text(self.delivery_method)}
+
+    def calculate_checkout_shipping(self, checkout_info, lines=None) -> TaxedMoney:
+        """Return checkout shipping price."""
+        # FIXME: Optimize checkout.is_shipping_required
+
+        shipping_method = self.delivery_method
+
+        if lines is not None and all(
+            isinstance(line, CheckoutLineInfo) for line in lines
+        ):
+            from .utils import is_shipping_required
+
+            shipping_required = is_shipping_required(lines)
+        else:
+            shipping_required = checkout_info.checkout.is_shipping_required()
+
+            if not shipping_method or not shipping_required:
+                return zero_taxed_money(checkout_info.checkout.currency)
+
+        shipping_price = shipping_method.channel_listings.get(
+            channel_id=checkout_info.checkout.channel_id,
+        ).get_total()
+
+        return quantize_price(
+            TaxedMoney(net=shipping_price, gross=shipping_price),
+            shipping_price.currency,
+        )
+
+    def is_valid_delivery_method(self) -> bool:
+        return bool(self.shipping_address)
+
+    def is_method_in_valid_methods(self, checkout_info: "CheckoutInfo") -> bool:
+        valid_delivery_methods = checkout_info.valid_delivery_methods
+        return bool(
+            valid_delivery_methods and self.delivery_method in valid_delivery_methods
+        )
+
+    def update_channel_listings(self, checkout_info: "CheckoutInfo") -> None:
+        checkout_info.shipping_method_channel_listings = (
+            ShippingMethodChannelListing.objects.filter(
+                shipping_method=self.delivery_method, channel=checkout_info.channel
+            ).first()
+        )
+
+
+@dataclass(frozen=True)
+class CollectionPointInfo(DeliveryMethodBase):
+    delivery_method: "Warehouse"
+    shipping_address: Optional["Address"]
+    order_key: str = "collection_point"
+
+    @property
+    def warehouse_pk(self):
+        return self.delivery_method.pk
+
+    @property
+    def is_local_collection_point(self):
+        return (
+            self.delivery_method.click_and_collect_option
+            == WarehouseClickAndCollectOption.LOCAL_STOCK
+        )
+
+    @property
+    def delivery_method_name(self) -> Dict[str, Optional[str]]:
+        return {"collection_point_name": smart_text(self.delivery_method)}
+
+    def get_warehouse_filter_lookup(self) -> Dict[str, Any]:
+        return (
+            {"warehouse__pk": self.delivery_method.pk}
```

Comment:
Maybe?
This should allow us to remove additional JOIN from SQL queries.
```suggestion
            {"warehouse_id": self.delivery_method.pk}
```

Reply:
👍  Done.